core/src/main/java/de/danoeh/antennapod/core/storage/DBReader.java

```diff
@@ -821,75 +823,28 @@ public static NavDrawerData getNavDrawerData() {
-                    return -1;
-                } else if (counterLhs == counterRhs) {
-                    return lhs.getTitle().compareToIgnoreCase(rhs.getTitle());
-                } else {
-                    return 1;
-                }
-            };
-        } else if (feedOrder == UserPreferences.FEED_ORDER_ALPHABETICAL) {
-            comparator = (lhs, rhs) -> {
-                String t1 = lhs.getTitle();
-                String t2 = rhs.getTitle();
-                if (t1 == null) {
-                    return 1;
-                } else if (t2 == null) {
-                    return -1;
-                } else {
-                    return t1.compareToIgnoreCase(t2);
-                }
-            };
-        } else if (feedOrder == UserPreferences.FEED_ORDER_MOST_PLAYED) {
-            final LongIntMap playedCounters = adapter.getPlayedEpisodesCounters();
-
-            comparator = (lhs, rhs) -> {
-                long counterLhs = playedCounters.get(lhs.getId());
-                long counterRhs = playedCounters.get(rhs.getId());
-                if (counterLhs > counterRhs) {
-                    // podcast with most played episodes first
-                    return -1;
-                } else if (counterLhs == counterRhs) {
-                    return lhs.getTitle().compareToIgnoreCase(rhs.getTitle());
-                } else {
-                    return 1;
-                }
-            };
-        } else {
-            final Map<Long, Long> recentPubDates = adapter.getMostRecentItemDates();
-            comparator = (lhs, rhs) -> {
-                long dateLhs = recentPubDates.containsKey(lhs.getId()) ? recentPubDates.get(lhs.getId()) : 0;
-                long dateRhs = recentPubDates.containsKey(rhs.getId()) ? recentPubDates.get(rhs.getId()) : 0;
-                return Long.compare(dateRhs, dateLhs);
-            };
-        }
 
-        Collections.sort(feeds, comparator);
         int queueSize = adapter.getQueueSize();
         int numNewItems = adapter.getNumberOfNewItems();
         int numDownloadedItems = adapter.getNumberOfDownloadedEpisodes();
 
+
         List<NavDrawerData.DrawerItem> items = new ArrayList<>();
         Map<String, NavDrawerData.FolderDrawerItem> folders = new HashMap<>();
         for (Feed feed : feeds) {
             for (String tag : feed.getPreferences().getTags()) {
+                long recentPubDate = recentPubDates.containsKey(feed.getId()) ? recentPubDates.get(feed.getId()) : -1;
                 NavDrawerData.FeedDrawerItem drawerItem = new NavDrawerData.FeedDrawerItem(feed, feed.getId(),
-                        feedCounters.get(feed.getId()));
-                if (FeedPreferences.TAG_ROOT.equals(tag)) {
-                    items.add(drawerItem);
-                    continue;
-                }
+                        feedCounters.get(feed.getId()), playedCounters.get(feed.getId(), -1), recentPubDate);
```

Comment:
This means that AntennaPod always needs to do 2 additional SQL queries that could potentially affect all episodes. Seems like it could cause performance problems. The app is rather sluggish with >150 subscriptions already.

Reply:
I was trying to extract the sorting logic to a sepearate class, but it was coupled with the adapter. I noticed that you had members in the NavDrawerData class to hold them. So I thought maybe will could store them in those, which does work. But didn't realize the performance overhead. The queries are rather complex.  

I guess I could pass in the adapter into the sorter class instead. Would that work for you?

Reply:
> I guess I could pass in the adapter into the sorter class instead. Would that work for you?

Hmm, not really happy with that either, to be honest. The database stuff shouldn't be passed over to the UI. How about introducing a `getFilteredSubscriptions` method to `DbReader` (that is also used internally by `getNavDrawerData`) that takes a `SubscriptionsFilter` as a parameter? That filter could then contain information about tags. Then the fragment does not need to do any filtering.